nexus/src/db/datastore.rs

```diff
@@ -4101,6 +4188,62 @@ mod test {
         logctx.cleanup_successful();
     }
 
+    #[tokio::test]
+    async fn test_service_upsert() {
+        let logctx = dev::test_setup_log("test_service_upsert");
+        let mut db = test_setup_database(&logctx.log).await;
+        let (opctx, datastore) = datastore_test(&logctx, &db).await;
+
+        // Create a sled on which the service should exist.
+        let sled_id = create_test_sled(&datastore).await;
+
+        // Create a new service to exist on this sled.
+        let service_id = Uuid::new_v4();
+        let addr = Ipv6Addr::LOCALHOST;
+        let kind = ServiceKind::Nexus;
+
+        let service = Service::new(service_id, sled_id, addr, kind);
+        let result =
+            datastore.service_upsert(&opctx, service.clone()).await.unwrap();
+        assert_eq!(service.id(), result.id());
+        assert_eq!(service.ip, result.ip);
+        assert_eq!(service.kind, result.kind);
+
+        db.cleanup().await.unwrap();
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_rack_initialize_is_idempotent() {
+        let logctx = dev::test_setup_log("test_rack_initialize_is_idempotent");
+        let mut db = test_setup_database(&logctx.log).await;
+        let (opctx, datastore) = datastore_test(&logctx, &db).await;
+
+        // Create a Rack, insert it into the DB.
+        let rack = Rack::new(Uuid::new_v4());
+        let result = datastore.rack_insert(&opctx, &rack).await.unwrap();
+        assert_eq!(result.id(), rack.id());
+        assert_eq!(result.initialized, false);
+
+        // Re-insert the Rack (check for idempotency).
+        datastore.rack_insert(&opctx, &rack).await.unwrap();
+        assert_eq!(result.id(), rack.id());
+        assert_eq!(result.initialized, false);
```

Comment:
These appear to be checking the `result` from L4224. Is that correct? It seems like they should check the result of the call on L4229.

Reply:
Good catch; fixed